libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayHistoryPlayer.kt

```diff
@@ -0,0 +1,88 @@
+
+import android.os.SystemClock
+import android.util.Log
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleOwner
+import kotlin.math.abs
+import kotlin.math.roundToInt
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+
+private typealias ReplayUpdatesListener = (ReplayUpdate) -> Unit
+
+/**
+ * This class is similar to a music player. It will include controls like play, pause, seek.
+ */
+class ReplayHistoryPlayer(
+    replayHistoryData: ReplayHistoryData
+) {
+    private val recordLookup = ReplayEventLookup(replayHistoryData)
+
+    private val recordUpdateListeners: MutableList<ReplayUpdatesListener> = mutableListOf()
+    private var mainJob: Job? = null
+
+    /**
+     * Events from the [ReplayHistoryData] will be published to your listener.
+     * Your subscriber will be removed after you call [finish]
+     */
+    fun observeReplayUpdates(function: ReplayUpdatesListener) {
+        recordUpdateListeners.add(function)
+    }
+
+    /**
+     * Meant to be called from an Android component, such as an Activity, Fragment, or ViewModel.
+     * This will begin playing the [ReplayHistoryData] and notifying listeners attached to
+     * [observeReplayUpdates]
+     */
+    fun play(lifecycleOwner: LifecycleOwner): Job {
+        Log.i("ReplayHistory", "Simulator started")
+
+        recordLookup.initPivot(timeSeconds())
+
+        val uiScope = CoroutineScope(Dispatchers.Main)
+        return uiScope.launch {
+            while (isActive && isSimulating(lifecycleOwner)) {
+                val loopStart = timeSeconds()
+
+                val recordUpdate = recordLookup.movePivot(timeSeconds())
+                if (recordUpdate.replayEventHappened.isNotEmpty()) {
+                    recordUpdateListeners.forEach { listener -> listener.invoke(recordUpdate) }
+                }
+
+                val loopElapsed = ((timeSeconds() - loopStart) * MILLIS_PER_SECOND).roundToInt()
+                val delayMillis = abs(replayUpdateSpeedMillis - loopElapsed)
+                delay(delayMillis)
+            }
+
+            Log.i("ReplayHistory", "Simulator ended")
+        }.also { this.mainJob = it }
+    }
+
+    /**
+     * Remove all the [observeReplayUpdates]. Designed to be called in tear down functions like
+     * Activity.onDestroy, Fragment.onDestroy, or ViewModel.onCleared
+     */
+    fun finish() {
+        mainJob?.cancel()
+        recordUpdateListeners.clear()
+    }
+
+    private fun isSimulating(lifecycleOwner: LifecycleOwner): Boolean {
+        return lifecycleOwner.lifecycle.currentState != Lifecycle.State.DESTROYED &&
+            !recordLookup.isComplete()
+    }
+
+    companion object {
+        // The frequency that replay updates will be broad-casted
+        private const val replayUpdateSpeedMillis = 1000L
```

Comment:
Should this be customizable (not hardcoded to 1 second) and somehow dynamically updated from `ReplayHistoryData`?

Reply:
didn't really want to expose this unless it proved to be useful to modify. also there is a way to make events emit at accurate times (described as the [future perfect simulator](https://github.com/mapbox/mapbox-navigation-android/pull/2689#discussion_r402623521)), which would make this interval obsolete 

1 second felt pretty good, it's pretty similar to the 1 second location frequency :)

Reply:
> 1 second felt pretty good, it's pretty similar to the 1 second location frequency :)

Sure! I was asking because that's going to change in the future 😬 But yeah, let's revisit in the future 😉